src/app/extensions/product-notifications/store/product-notification/product-notification.reducer.ts

```diff
@@ -0,0 +1,67 @@
+import { EntityState, createEntityAdapter } from '@ngrx/entity';
+import { createReducer, on } from '@ngrx/store';
+
+import { HttpError } from 'ish-core/models/http-error/http-error.model';
+import { setErrorOn, setLoadingOn, unsetLoadingAndErrorOn } from 'ish-core/utils/ngrx-creators';
+
+import { ProductNotification } from '../../models/product-notification/product-notification.model';
+
+import { productNotificationsActions, productNotificationsApiActions } from './product-notification.actions';
+
+export const productNotificationAdapter = createEntityAdapter<ProductNotification>();
+
+export interface ProductNotificationState extends EntityState<ProductNotification> {
+  loading: boolean;
+  error: HttpError;
+}
+
+export const initialState: ProductNotificationState = productNotificationAdapter.getInitialState({
+  loading: false,
+  error: undefined,
+});
+
+export const productNotificationReducer = createReducer(
+  initialState,
+  setLoadingOn(
+    productNotificationsActions.loadProductNotifications,
+    productNotificationsActions.createProductNotification,
+    productNotificationsActions.updateProductNotification,
+    productNotificationsActions.deleteProductNotification
+  ),
+  setErrorOn(
+    productNotificationsApiActions.loadProductNotificationsFail,
+    productNotificationsApiActions.createProductNotificationFail,
+    productNotificationsApiActions.updateProductNotificationFail,
+    productNotificationsApiActions.deleteProductNotificationFail
+  ),
+  unsetLoadingAndErrorOn(
+    productNotificationsApiActions.loadProductNotificationsSuccess,
+    productNotificationsApiActions.createProductNotificationSuccess,
+    productNotificationsApiActions.updateProductNotificationSuccess,
+    productNotificationsApiActions.deleteProductNotificationSuccess
+  ),
+  on(productNotificationsApiActions.loadProductNotificationsSuccess, (state, action) =>
+    /**
+     * Product notifications can be deleted on server side when the notification requirements
+     * are met and the notification email was sent. Therefore, all product notifications
+     * have to be removed from the state which are not returned from the service before they
+     * are loaded, displayed or used.
+     */
+    productNotificationAdapter.upsertMany(action.payload.productNotifications, {
```

Comment:
why don't you work with addAll here? It replaces the current collection with the provided collection and this is what you want to do here?

Reply:
The comment is updated. If we use setAll, the notification list would always be empty at first and filled when the REST request is done. With `upsertMany` / `removeMany` we show the notifications in the state and update the list (and remove what may be deleted on server side).